crates/server/src/store/blocklist/authority.rs

```diff
@@ -0,0 +1,430 @@
+/// of this configuration.
+///
+/// The blocklist authority also supports the consult interface, which allows an authority to review
+/// a query/response that has been processed by another authority, and, optionally, overwrite that
+/// response before returning it to the requestor.  There is an example of this configuration in
+/// tests/test-data/test_configs/example_consulting_blocklist.toml.  The main intended use of this
+/// feature is to allow log-only configurations, to allow administrators to see if blocklist domains
+/// are being queried.  While this can be configured to overwrite responses, it is not recommended
+/// to do so - it is both more efficient, and more secure, to allow the blocklist to drop queries
+/// pre-emptively, as in the first example.
+///
+/// TODO:
+///  * Add (optional) support for logging the client IP address.  This will require some Authority
+///    trait changes to accomplish.
+///  * Add support for a block-list block-list: allow the user to configure a list of patterns that
+///    will never be added to a blocklist (such as their own domain.)
+///  * Add support for regex matching
+pub struct BlocklistAuthority {
+    origin: LowerName,
+    blocklist: HashMap<LowerName, bool>,
+    wildcard_match: bool,
+    min_wildcard_depth: u8,
+    sinkhole_ip: Ipv4Addr,
+    consult_action: BlocklistConsultAction,
+}
+
+impl BlocklistAuthority {
+    /// Read the Authority for the origin from the specified configuration
+    pub async fn try_from_config(
+        origin: Name,
+        _zone_type: ZoneType,
+        config: &BlocklistConfig,
+        root_dir: Option<&Path>,
+    ) -> Result<Self, String> {
+        info!("loading blocklist config: {origin}");
+
+        let sinkhole_ip = Ipv4Addr::from_str(&config.sinkhole_ip[..]).unwrap(/* Panic on startup on an invalid IP */);
+
+        let mut authority = Self {
+            origin: origin.into(),
+            blocklist: HashMap::new(),
+            wildcard_match: config.wildcard_match,
+            min_wildcard_depth: config.min_wildcard_depth,
+            sinkhole_ip,
+            consult_action: config.consult_action,
+        };
+
+        // Load block lists into the block table cache for this authority.
+        for bl in &config.lists {
+            info!("Adding blocklist {bl}");
+            authority
+                .add(format!(
+                    "{}/{bl}",
+                    root_dir
+                             .unwrap(/* Panic on startup on an invalid path */)
+                             .display()
+                ))
+                .await;
+        }
+
+        Ok(authority)
+    }
+
+    /// Add a configured block list to the in-memory cache.
+    pub async fn add(&mut self, file: String) -> bool {
+        /* Panic on an invalid list file */
+        let mut handle =
+            File::open(&file).unwrap_or_else(|_| panic!("unable to open block list file '{file}'"));
```

Comment:
IMO we should not panic for this, make this fallible instead?

Reply:
We panic if a primary zone is configured but the zone file can't be read. I think this is essentially the same thing.

Reply:
I don't disagree but IMO that design is just faulty, too. Parts of the Hickory DNS code base are pretty old or haven't kept up with best practices. Especially for libraries, we should avoid panicking.